misc/python/materialize/output_consistency/operations/operation.py

```diff
@@ -0,0 +1,94 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from typing import Optional
+
+from materialize.output_consistency.data_type.data_type_group import DataTypeGroup
+from materialize.output_consistency.operations.operation_args_validator import (
+    OperationArgsValidator,
+)
+from materialize.output_consistency.operations.operation_param import OperationParam
+
+
+class OperationWithNParams:
+    def __init__(
+        self,
+        pattern: str,
+        params: list[OperationParam],
+        return_type_group: DataTypeGroup,
+        args_validators: Optional[set[OperationArgsValidator]] = None,
+        commutative: bool = False,
+    ):
+        if args_validators is None:
+            args_validators = set()
+
+        self.pattern = pattern
+        self.params = params
+        self.param_count = len(params)
+        self.return_type_group = return_type_group
+        self.commutative = commutative
+        self.args_validators: set[OperationArgsValidator] = args_validators
+
+
+class OperationWithOneParam(OperationWithNParams):
+    def __init__(
+        self,
+        pattern: str,
+        param: OperationParam,
+        return_type_group: DataTypeGroup,
+        args_validators: Optional[set[OperationArgsValidator]] = None,
+    ):
+        self.pattern = pattern
+        super().__init__(pattern, [param], return_type_group, args_validators, False)
+
+
+class OperationWithTwoParams(OperationWithNParams):
+    def __init__(
+        self,
+        pattern: str,
+        param1: OperationParam,
+        param2: OperationParam,
+        return_type_group: DataTypeGroup,
+        args_validators: Optional[set[OperationArgsValidator]] = None,
+        commutative: bool = False,
+    ):
+        self.pattern = pattern
+        super().__init__(
+            pattern, [param1, param2], return_type_group, args_validators, commutative
+        )
+
+
+class UnaryFunction(OperationWithOneParam):
```

Comment:
How is `UnaryFunction` different from `OperationWithOneParam`? If they are practically the same, consider having just a single class with `Function` in its name.